cw-dex/tests/osmosis_tests.rs

```diff
@@ -0,0 +1,378 @@
+    fn test_withdraw_liquidity(pool_type: OsmosisPoolType, initial_liquidity: Vec<u64>) {
+        let (runner, accs, pool_id, contract_addr) =
+            setup_pool_and_contract(pool_type, initial_liquidity, None).unwrap();
+        let admin = &accs[0];
+        let lp_token_denom = format!("gamm/pool/{}", pool_id);
+
+        //Query admin LP token balance
+        let admin_lp_token_balance =
+            bank_balance_query(&runner, admin.address(), lp_token_denom.clone()).unwrap();
+
+        // Send LP tokens to the test contract
+        bank_send(
+            &runner,
+            admin,
+            &contract_addr,
+            vec![Coin::new(
+                admin_lp_token_balance.u128(),
+                lp_token_denom.clone(),
+            )],
+        )
+        .unwrap();
+
+        // Withdraw liquidity. We are not allowed to withdraw all liquidity on osmosis.
+        let withdraw_msg = ExecuteMsg::WithdrawLiquidity {
+            amount: admin_lp_token_balance.checked_sub(Uint128::one()).unwrap(),
+        };
+        runner
+            .execute_cosmos_msgs::<MsgExecuteContractResponse>(
+                &[withdraw_msg.into_cosmos_msg(contract_addr.clone(), vec![])],
+                admin,
+            )
+            .unwrap();
+
+        // Query LP token balance after
+        let lp_token_balance_after =
+            bank_balance_query(&runner, contract_addr, lp_token_denom).unwrap();
+
+        // Assert that LP token balance is 1
+        assert_eq!(lp_token_balance_after, Uint128::one());
+    }
+
+    fn stake_all_lp_tokens<'a, R: Runner<'a>>(
+        runner: &'a R,
+        contract_addr: String,
+        pool_id: u64,
+        signer: &SigningAccount,
+    ) -> ExecuteResponse<MsgExecuteContractResponse> {
+        // Query LP token balance
+        let lp_token_balance = bank_balance_query(
+            runner,
+            contract_addr.clone(),
+            format!("gamm/pool/{}", pool_id),
+        )
+        .unwrap();
+
+        // Stake LP tokens
+        let stake_msg = ExecuteMsg::Stake {
+            amount: lp_token_balance,
+        };
+        runner
+            .execute_cosmos_msgs::<MsgExecuteContractResponse>(
+                &[stake_msg.into_cosmos_msg(contract_addr, vec![])],
+                signer,
+            )
+            .unwrap()
+    }
+
+    #[test_case(86_400 ; "one day lock")]
+    #[test_case(604_800 ; "one week lock")]
+    #[test_case(1_209_600 ; "two week lock")]
+    #[test_case(1 => matches Err(_) ; "invalid lock duration")]
+    fn test_stake_and_unlock(lock_duration: u64) -> RunnerResult<()> {
+        let (runner, accs, pool_id, contract_addr) = setup_pool_and_contract(
+            OsmosisPoolType::Basic,
+            vec![1_000_000, 1_000_000],
+            Some(lock_duration),
+        )?;
+        let admin = &accs[0];
+
+        // Query LP token balance
+        let lp_token_denom = format!("gamm/pool/{}", pool_id);
+        let lp_token_balance =
```

Comment:
Consider renaming this variable as its used later below and it's there a bit unclear which balance it is.